lib/pure/times.nim

```diff
@@ -172,30 +178,43 @@ proc getDaysInYear*(year: int): int =
   ## Get the number of days in a ``year``
   result = 365 + (if isLeapYear(year): 1 else: 0)
 
-proc assertValidDate(monthday: MonthdayRange, month: Month, year: int) {.inline.} =
-  assert monthday <= getDaysInMonth(month, year),
-    $year & "-" & $ord(month) & "-" & $monthday & " is not a valid date"
+proc assertValidDate(d: Date) {.inline.} =
+  assert d.monthday <= getDaysInMonth(d.month, d.year),
+    $d.year & "-" & $ord(d.month) & "-" & $d.monthday & " is not a valid date"
+
+proc toEpochDate*(time: Time): EpochDate =
+  EpochDate(time.int64 div secondsInDay)
+
+proc toEpochDate*(epoch_days: int32): EpochDate =
+  epoch_days.EpochDate
+
+proc toEpochDate(year, month, day: int): EpochDate =
+  ## Get the epoch date from a year/month/day combination
+  ## Note that numbers can be positive and negative
+  ## as well as outside of valid month, monthday range. 
+  ## Example: 2012-(-2)-32 is treated as 2011-12-02
+  ## This feature is used for efficient interval calculation
 
-proc toEpochDay(monthday: MonthdayRange, month: Month, year: int): int64 =
-  ## Get the epoch day from a year/month/day date.
-  ## The epoch day is the number of days since 1970/01/01 (it might be negative).
-  assertValidDate monthday, month, year
   # Based on http://howardhinnant.github.io/date_algorithms.html
-  var (y, m, d) = (year, ord(month), monthday.int)
-  if m <= 2:
+  var (y, m, d) = (year, month, day)
+  while m <= 2:
     y.dec
+    m.inc 12
 
   let era = (if y >= 0: y else: y-399) div 400
   let yoe = y - era * 400
-  let doy = (153 * (m + (if m > 2: -3 else: 9)) + 2) div 5 + d-1
+  let doy = (153 * (m - 3) + 2) div 5 + d-1
   let doe = yoe * 365 + yoe div 4 - yoe div 100 + doy
-  return era * 146097 + doe - 719468
+  EpochDate(era * 146097 + doe - 719468)
 
-proc fromEpochDay(epochday: int64): tuple[monthday: MonthdayRange, month: Month, year: int] =
+proc toEpochDate*(d: Date): EpochDate =
+  assertValidDate d
+  toEpochDate(d.year, d.month.int, d.monthday.int)
+
+proc date*(epochdate: EpochDate): Date =
```

Comment:
I dislike the naming of this. I think `fromEpochDate` better matches the rest of the API in the `times` module. If that is to ambiguous (since it's not clear what it returns, but I think it's fine), then `toDate` is also fine.

Reply:
done for now, unless idea of single entry `date`  will be supported